[3.3] 1924 export associations as resource relationship

Atticus29 opened the pull request:
# Description

This PR allows for the download of related records. It addresses issue #1924 .

From the issue specs:

> Adds DwcArchiverResourceRelationship class (following format similar to DwcArchiverIdentifier, DwcArchiverMaterialSample, and DwcArchiverAttribute
Add include Resource Relationship extension checkbox to all download interfaces and web services
Currently External and Internal Associations are concatenated into associatedOccurrences within occurrences.csv core file.
associatedOccurrences will be removed from DwC-Archive export (simply comment out lines 137 and 138 within DwcArchiverOccurrence). We might remove this column from the omoccurrences schema, but let's just comment it out for now.
Within DwcArchiverOccurrence getAssociationStr and getAssociationJSON will be completely removed.
appendSpecimenDuplicateAssociations function will be removed with functionality integrated as a batch function, similar to functions called between lines 1821 and 1827 within DwcArchiverCore
Currently observation associations are appended into the associatedTaxa field, which needs to be removed. The associatedTaxa field will remain in the DwC Archive export, but should now only contain verbatim text from the associatedTaxa field.

# QA Notes

Live on dev001. In addition to all of the nuances of what we should map from Symbiota Native fields to DwC, note that I haven't tested the Darwin Core-only download or skipping the download entirely in a while.

# Notes/Known Issues

This PR does not accommodate external or observational associations. They are in a [separate issue](https://github.com/BioKIC/Symbiota/issues/2091).

# Pull Request Checklist:

# Pre-Approval

- [x] There is a description section in the pull request that details what the proposed changes do. It can be very brief if need be, but it ought to exist.
- [ ] Hotfixes should be branched off of the `master` branch and PR'd using the **merge** option (not squashed) into the `hotfix` branch.
    - N/A
- [x] Features and backlog bugs should be merged into the `Development` branch, **NOT** `master`
- [ ] All new text is preferably internationalized (i.e., no end-user-visible text is hard-coded on the PHP pages), and the [spreadsheet tracking internationalizations](https://docs.google.com/spreadsheets/d/133fps9w2pUCEjUA6IGCcQotk7dn9KvepMXJ2IWUZsE8/edit?usp=sharing) has been updated either with a new row or with checkmarks to existing rows.
    - N/A
- [x] There are no linter errors
- [ ] New features have responsive design (i.e., look aesthetically pleasing both full screen and with small or mobile screens)
    - N/A
- [x] [Symbiota coding standards](https://docs.google.com/document/d/1-FwCZP5Zu4f-bPwsKeVVsZErytALOJyA2szjbfSUjmc/edit?usp=sharing) have been followed
- [x] If any files have been reformatted (e.g., by an autoformatter), the reformat is its own, separate commit in the PR
    - If so, noted in code diff
- [x] Comment which GitHub issue(s), if any does this PR address
    - [https://github.com/BioKIC/Symbiota/issues/1924](https://github.com/BioKIC/Symbiota/issues/1924)
- [ ] If this PR makes any changes that would require additional configuration of any Symbiota portals outside of the files tracked in this repository, make sure that those changes are detailed in [this document](https://docs.google.com/document/d/1T7xbXEf2bjjm-PMrlXpUBa69aTMAIROPXVqJqa2ow_I/edit?usp=sharing).
    - N/A

# Post-Approval

- [ ] It is the code author's responsibility to merge their own pull request after it has been approved
- [ ] If this PR represents a merge into the `Development` branch, remember to use the **squash & merge** option
- [ ] If this PR represents a merge into the `hotfix` branch, remember to use the **merge** option (i.e., no squash).
- [ ] If this PR represents a merge from the `Development` branch into the master branch, remember to use the **merge** option
- [ ] If this PR represents a merge from the `hotfix` branch into the `master` branch use the **squash & merge** option
  - [ ] a subsequent PR from `master` into `Development` should be made with the **merge** option (i.e., no squash).
  - [ ] **Immediately** delete the `hotfix` branch and create a new `hotfix` branch
  - [ ] increment the Symbiota version number in the symbase.php file and commit to the `hotfix` branch.
- [ ] If the dev team has agreed that this PR represents the last PR going into the `Development` branch before a tagged release (i.e., before an imminent merge into the master branch), make sure to notify the team and [lock the `Development` branch](https://github.com/BioKIC/Symbiota/settings/branches) to prevent accidental merges while QA takes place. Follow the release protocol [here](https://github.com/BioKIC/Symbiota/blob/master/docs/release-protocol.md).
- [ ] Don't forget to delete your feature branch upon merge. Ignore this step as required.

Thanks for contributing and keeping it clean!

Atticus29 committed:
create DwcArchiverAssociation.php and implement mapping in setFieldArr method

Atticus29 committed:
finish implementing first version of classes/DwcArchiverAssociation.php

Atticus29 committed:
silence association sql print

Atticus29 committed:
add translations, include Resource Relationship extension checkbox, and begin to implement some of the export logic

Atticus29 committed:
fix minor bug

Atticus29 committed:
make association download work. Populate? No.

Atticus29 committed:
allow writeOutRecordBlock to handle the union required for two-way association search

Atticus29 committed:
add symbiota native columns to the export

themerekat commented:
Change verbiage here from "include Associated Records" to "include Resource Relationships (associations & linked resources)"
![image](https://github.com/user-attachments/assets/2f858685-4a38-4538-80a5-86501d7828db)

Atticus29 committed:
make it so there is no data cutoff for xdebugger

Atticus29 committed:
fix bug where download was not working when association search was being used

Atticus29 committed:
comment out associatedOccurrences

Atticus29 commented on .vscode/launch.json:
```diff
@@ -1,48 +1,51 @@
+      "type": "php",
+      "request": "launch",
+      "port": 9003,
+      "xdebugSettings": {
```
This file was autoformatted. But also the xdebugSettings section was added because my xDebugger values were being very confusingly truncated. Fixed here slash not at all sure why this is being tracked in the repo.

Atticus29 committed:
remove appendSpecimenDuplicateAssociations, getAssociationJSON, and getAssociationStr

Atticus29 committed:
remove some cruft and change some translations

Atticus29 commented:
> include Resource Relationships (associations & linked resources)

Fixed in 20a3b08cc.

Atticus29 committed:
fix bug where occurrence download was not respecting association constraints

Atticus29 committed:
fix bug where ->associationManager->getAssociatedRecords was being run twice

Atticus29 committed:
begin to implement the changes where omoccurassociations is joined to omoccurrences so that occurrenceID can be queried

egbot committed:
association export mod

- Load associations via 2 separate queries:
-- one using occid
-- the other using occidAssociate with the relationship translated to its inverse

egbot committed:
Merge pull request #1961 from BioKIC/associations-export-egbot

association export mod

Atticus29 committed:
cherry pick copy changes and resolve merge conflicts

Atticus29 committed:
resolve merge conflicts with Development

Atticus29 committed:
add warning to users about uniqueness of occurrenceId over other identifiers

Atticus29 committed:
fix typo in recordID

Atticus29 committed:
add temporary columns to aid in troubleshooting

Atticus29 committed:
fix typos and minor bugs in collections/search/js/searchform.js

Atticus29 committed:
resolve merge conflicts with Development

Atticus29 committed:
fix bug where  was not targeting the correct keys for DWC-formatted download

Atticus29 committed:
rename DwcArchiverAssociation to DwcArchiverResourceRelationship

Atticus29 committed:
remove renamed class file

Atticus29 committed:
comment out appendSpecimenDuplicateAssociations call per ticket issues recommendation

Atticus29 committed:
remove no-longer-needed schema change

Atticus29 committed:
remove unnecessary TODO

Atticus29 committed:
get more things from the omoccurrences table. Fix bug in setSqlBase logic

Atticus29 committed:
add left join of the object omoccurrences entry to fetch relatedResourceID, which required some logic change since omoccurrences does not have objectID nor instanceID

Atticus29 committed:
refine relatedResourceID and resourceRelationshipID

Atticus29 committed:
make relatedResourceID refer to the object

Atticus29 committed:
continue improvements to resourceRelationshipID and relatedResourceID

Atticus29 commented on classes/DwcArchiverResourceRelationship.php:
```diff
@@ -0,0 +1,134 @@
+		$termArr['deleteMeOccidAssociate'] = 'https://dwc.tdwg.org/terms/#dwc:resourceRelationshipID';
+		$columnArr['deleteMeOccidAssociate'] = 'oa.occidAssociate';
+		$termArr['resourceRelationshipID'] = 'https://dwc.tdwg.org/terms/#dwc:resourceRelationshipID';
+		// $columnArr['resourceRelationshipID'] = 'IFNULL(IFNULL(IFNULL(oa.objectID, o.occurrenceID), o.recordID), oa.resourceUrl)';
```
I'll delete some of these commented-out lines once we're more confident about the current strategy.

Atticus29 commented on classes/DwcArchiverResourceRelationship.php:
```diff
@@ -0,0 +1,134 @@
+		return array_diff_key($dataArr, array_flip($trimArr));
+	}
+
+	// @TODO decide if setDynamicFields is needed
```
I don't think so, right?

Atticus29 commented on classes/DwcArchiverResourceRelationship.php:
```diff
@@ -0,0 +1,134 @@
+				foreach($this->fieldArr['fields'] as $colName){
+					if($colName) $sqlFrag .= ', ' . $colName;
+				}
+				// $this->sqlBase = 'SELECT ' . trim($sqlFrag, ', ') . ' FROM omoccurassociations ';
```
Will remove upon approval.

Atticus29 commented on collections/admin/importextended.php:
```diff
@@ -337,11 +338,14 @@ function importTypeChanged(selectElement){
 								<input name="collid" type="hidden" value="<?= $collid ?>" >
 								<input name="MAX_FILE_SIZE" type="hidden" value="10000000" />
-								<button name="submitAction" type="submit" value="initiateImport"><?= $LANG['INITIALIZE_IMPORT'] ?></button>
+								<button name="submitAction" class="submit__button--no-left-margin" type="submit" value="initiateImport"><?= $LANG['INITIALIZE_IMPORT'] ?></button>
```
Little styling fix

Atticus29 commented on collections/datasets/datapublisher.php:
```diff
@@ -8,7 +8,7 @@
 header('Content-Type: text/html; charset=' . $CHARSET);
 
-$collid = array_key_exists('collid', $_REQUEST) ? filter_var($_REQUEST['collid'], FILTER_SANITIZE_NUMBER_INT) : 0;
+$collid = array_key_exists('collid', $_REQUEST) ? filter_var($_REQUEST['collid'], FILTER_SANITIZE_NUMBER_INT) : 0; // @TODO collid is really coming from db in the searchvar attribute of the request, right? So it'll always be incorrect here?
```
Maybe a future easy issue to clean up?

Atticus29 commented on collections/admin/importextended.php:
```diff
@@ -337,11 +338,14 @@ function importTypeChanged(selectElement){
 									<option value="observational"><?= $LANG['OBSERVATION'] ?></option>
 								</select>
+								<div class="top-breathing-room-rel danger" style="color: var(--danger-color);">
+									<caption><?= $LANG['ASSOCIATION_UPLOAD_WARNING'] ?></caption>
```
A little behavioral nudge for the end users who could footgun themselves with bulk uploading associations matching to the correct catalogNumber in an off-target collection (since cross-collection associations are correctly allowed as well).

Atticus29 commented on content/lang/collections/admin/importextended.en.php:
```diff
@@ -33,7 +33,7 @@
 $LANG['BATCH_DELETE'] = 'Batch Delete';
 $LANG['NEW_BLANK_RECORD'] = 'Link media to new blank record if catalog number does not exist';
-$LANG['MATCHING_IDENTIFIERS'] = 'Update or delete records with matching identifiers (subject identifier plus objectID or resourceUrl)';
+$LANG['MATCHING_IDENTIFIERS'] = 'Update or delete existing association with matching identifiers (subject identifier plus objectID or resourceUrl)';
```
Per conversations with @themerekat , minor copy adjustments here.

Atticus29 committed:
add missing translations

themerekat commented:
I think something is going slightly wrong with the exporter's verbatimScientificName (not sure if this is upon export or during the creation of the association). The "hostOf" relationship should be with Apidae

From the export:
![image](https://github.com/user-attachments/assets/ede5c21a-248a-49d4-9b8e-054f6c7247d4)

From the public display:
![image](https://github.com/user-attachments/assets/33d7ae50-e7fc-4fce-b510-5bc28fbef1fc)

Atticus29 commented:
> I think something is going slightly wrong with the exporter's verbatimScientificName (not sure if this is upon export or during the creation of the association). The "hostOf" relationship should be with Apidae
> 
> From the export: ![image](https://private-user-images.githubusercontent.com/16688784/405810474-ede5c21a-248a-49d4-9b8e-054f6c7247d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc1ODU2MTcsIm5iZiI6MTczNzU4NTMxNywicGF0aCI6Ii8xNjY4ODc4NC80MDU4MTA0NzQtZWRlNWMyMWEtMjQ4YS00OWQ0LTliOGUtMDU0ZjZjNzI0N2Q0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIyVDIyMzUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlZTJiZmI2MTUxMGZjYzU2ZmVhMWE1ZWE2YTY5ZTY2YWI0NTU4ODllOTVhMDUyMjViOWRlNmE1ZmM1ZjI3ZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.wqBz23bH02b0xSArmJ0LXhTAMlaG35447uOPzuvyWCQ)
> 
> From the public display: ![image](https://private-user-images.githubusercontent.com/16688784/405810788-33d7ae50-e7fc-4fce-b510-5bc28fbef1fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc1ODU2MTcsIm5iZiI6MTczNzU4NTMxNywicGF0aCI6Ii8xNjY4ODc4NC80MDU4MTA3ODgtMzNkN2FlNTAtZTdmYy00ZmNlLWI1MTAtNWJjMjhmYmVmMWZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIyVDIyMzUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY0OGU1NjBkZmVlNGI2ZGVjOTE0MmQzOWJiNGM0NjdiNzNiM2ZjNGE3NjU1ZDMzNWZjMTdhZDI3NzQ4ZDk5MzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZAZ8WVGuI3E0eO7ErXdu-YABj_Ml57Mh4Cnd6cE0Exo)

Oh interesting. This is an observational association. Maybe we make that a problem for FutureMark?

Atticus29 committed:
resolve merge conflicts

Atticus29 committed:
1924 pt2 external and observational occurrences (#2135)

* swap the occid and occidAssociate in getAssociatedRecords in classes/AssociationManager.php in order to get any results in the external observation export

* add third logic block to getAssociatedRecords targetting observations and external observations

* begin to remove cruft on 1924-pt2-external-and-observational-occurrences branch

* remove more cruft

* remove more cruft

* respond to code review feedback

Atticus29 committed:
resolve merge conflicts with Development

themerekat reviewed (changes requested):
Taxon observations don't seem to be coming out right. Here is a record:
![image](https://github.com/user-attachments/assets/553b5716-acd9-4215-9680-9206e740ddb1)

And here are the resulting associations:
![image](https://github.com/user-attachments/assets/b480b2e8-ef6c-4d51-9ffe-500178616dd3)

Notice that they all have the same scientific name! I think that the verbatimSciName value is coming from the wrong table.

themerekat commented:
If the relationshipType value is "derivedFromSameIndividual", does it automatically inherit the verbatimSciName from the occurrence record? That's what I'm seeing for records like this: 
![image](https://github.com/user-attachments/assets/e2d3907f-0da9-4981-bba1-fab36ad6c37d)

(I think that's fine, just want to make sure that this isn't happening for ALL relationship types)

themerekat commented:
> If the relationshipType value is "derivedFromSameIndividual", does it automatically inherit the verbatimSciName from the occurrence record? That's what I'm seeing for records like this: ![image](https://private-user-images.githubusercontent.com/16688784/432390510-e2d3907f-0da9-4981-bba1-fab36ad6c37d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQzMDIyMjcsIm5iZiI6MTc0NDMwMTkyNywicGF0aCI6Ii8xNjY4ODc4NC80MzIzOTA1MTAtZTJkMzkwN2YtMGRhOS00OTgxLWJiYTEtZmFiMzZhZDZjMzdkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDEwVDE2MTg0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExZmU2MzIwOGFlYThjYmY2MDQ3NWVjYjdjMWFlMjhlNTkyNmMxNWM4YjcxZjA1ZmRjYmEyZDA5MjQ4YmE4OTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mfEr1FmWW_idY719r8dbwhvravg0ixrcQoVmQVOpmZY)
> 
> (I think that's fine, just want to make sure that this isn't happening for ALL relationship types)

Did some testing, and this is not the case. Even if I change the "relationshipType" to "hostOf", the verbatimsciname remains the same. This is not correct, because the verbatimSciName of the association should be the one in the linked record, not the Symbiota occurrence. I think this is the same issue as I mentioned previously with taxon observations.

Atticus29 committed:
resolve merge conflicts

Atticus29 commented:
> > If the relationshipType value is "derivedFromSameIndividual", does it automatically inherit the verbatimSciName from the occurrence record? That's what I'm seeing for records like this: ![image](https://private-user-images.githubusercontent.com/16688784/432390510-e2d3907f-0da9-4981-bba1-fab36ad6c37d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQzMDIyMjcsIm5iZiI6MTc0NDMwMTkyNywicGF0aCI6Ii8xNjY4ODc4NC80MzIzOTA1MTAtZTJkMzkwN2YtMGRhOS00OTgxLWJiYTEtZmFiMzZhZDZjMzdkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDEwVDE2MTg0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExZmU2MzIwOGFlYThjYmY2MDQ3NWVjYjdjMWFlMjhlNTkyNmMxNWM4YjcxZjA1ZmRjYmEyZDA5MjQ4YmE4OTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mfEr1FmWW_idY719r8dbwhvravg0ixrcQoVmQVOpmZY)
> > (I think that's fine, just want to make sure that this isn't happening for ALL relationship types)
> 
> Did some testing, and this is not the case. Even if I change the "relationshipType" to "hostOf", the verbatimsciname remains the same. This is not correct, because the verbatimSciName of the association should be the one in the linked record, not the Symbiota occurrence. I think this is the same issue as I mentioned previously with taxon observations.

@themerekat after merging in some of the recent work from other efforts (e.g., the observational association search fix), is this working any better now, or are you still experiencing the issue?

themerekat commented:
@Atticus29 , well I tried to download the results from this query
![image](https://github.com/user-attachments/assets/6796ba6e-28be-493e-a81c-4e057960f01d)

and it sounded like my computer was trying to take off (and the download never happened), so I wonder if something is stuck/looping with these criteria.

Atticus29 committed:
fix association download bug

Atticus29 commented:
> @Atticus29 , well I tried to download the results from this query ![image](https://private-user-images.githubusercontent.com/16688784/434535648-6796ba6e-28be-493e-a81c-4e057960f01d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUzNDkwNjUsIm5iZiI6MTc0NTM0ODc2NSwicGF0aCI6Ii8xNjY4ODc4NC80MzQ1MzU2NDgtNjc5NmJhNmUtMjhiZS00OTNlLWE4MWMtNGUwNTc5NjBmMDFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIyVDE5MDYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZhNGU5ZjdmOTAzZDJkNGFhN2EyMzJmYTkyZjhkMzRmZWE1MTEzNDg4NmE3N2Y3ZDYyYWVhOGUyNWMzMzkzMmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xIywjteISyzxJktdd6pSDpW3KAmE1Aym2zLCElgg0Cg)
> 
> and it sounded like my computer was trying to take off (and the download never happened), so I wonder if something is stuck/looping with these criteria.

Fixed in e50b9eb78. Still working on the sciName mis-match mentioned above.

Atticus29 committed:
add taxa join to get sciName

Atticus29 committed:
clean up

Atticus29 committed:
clean up cruft

Atticus29 commented:
> > If the relationshipType value is "derivedFromSameIndividual", does it automatically inherit the verbatimSciName from the occurrence record? That's what I'm seeing for records like this: ![image](https://private-user-images.githubusercontent.com/16688784/432390510-e2d3907f-0da9-4981-bba1-fab36ad6c37d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQzMDIyMjcsIm5iZiI6MTc0NDMwMTkyNywicGF0aCI6Ii8xNjY4ODc4NC80MzIzOTA1MTAtZTJkMzkwN2YtMGRhOS00OTgxLWJiYTEtZmFiMzZhZDZjMzdkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDEwVDE2MTg0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExZmU2MzIwOGFlYThjYmY2MDQ3NWVjYjdjMWFlMjhlNTkyNmMxNWM4YjcxZjA1ZmRjYmEyZDA5MjQ4YmE4OTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mfEr1FmWW_idY719r8dbwhvravg0ixrcQoVmQVOpmZY)
> > (I think that's fine, just want to make sure that this isn't happening for ALL relationship types)
> 
> Did some testing, and this is not the case. Even if I change the "relationshipType" to "hostOf", the verbatimsciname remains the same. This is not correct, because the verbatimSciName of the association should be the one in the linked record, not the Symbiota occurrence. I think this is the same issue as I mentioned previously with taxon observations.

Should be fixed in 33cf51f97.

themerekat reviewed (changes requested):
Rats, something is still not correct! verbatimSciName should always be the scientific name of the **object**, not the subject. I found a case of ecologicallyOccursWith where this is correct:

![image](https://github.com/user-attachments/assets/80399bf1-59ab-46a4-ae99-e27876f70969)
Subject = Abronia maritima, Object = Isocoma menziesii var. vernonioides

However, the inverse is not correct

![image](https://github.com/user-attachments/assets/35905269-70bd-44d9-814d-2a7e286f33ce)

This is saying that subject = Isocoma menziesii var. vernonioides AND object = Isocoma menziesii var. vernonioides.

themerekat commented:
> > @Atticus29 , well I tried to download the results from this query ![image](https://private-user-images.githubusercontent.com/16688784/434535648-6796ba6e-28be-493e-a81c-4e057960f01d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUzNDkwNjUsIm5iZiI6MTc0NTM0ODc2NSwicGF0aCI6Ii8xNjY4ODc4NC80MzQ1MzU2NDgtNjc5NmJhNmUtMjhiZS00OTNlLWE4MWMtNGUwNTc5NjBmMDFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIyVDE5MDYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZhNGU5ZjdmOTAzZDJkNGFhN2EyMzJmYTkyZjhkMzRmZWE1MTEzNDg4NmE3N2Y3ZDYyYWVhOGUyNWMzMzkzMmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xIywjteISyzxJktdd6pSDpW3KAmE1Aym2zLCElgg0Cg)
> > and it sounded like my computer was trying to take off (and the download never happened), so I wonder if something is stuck/looping with these criteria.
> 
> Fixed in [e50b9eb](https://github.com/Symbiota/Symbiota/commit/e50b9eb7867f6cf8c9f9a3b903830ab44af13ba2). Still working on the sciName mis-match mentioned above.

This is fixed!

themerekat commented:
> Rats, something is still not correct! verbatimSciName should always be the scientific name of the **object**, not the subject. I found a case of ecologicallyOccursWith where this is correct:
> 
> ![image](https://private-user-images.githubusercontent.com/16688784/439698081-80399bf1-59ab-46a4-ae99-e27876f70969.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYxMjc1MzEsIm5iZiI6MTc0NjEyNzIzMSwicGF0aCI6Ii8xNjY4ODc4NC80Mzk2OTgwODEtODAzOTliZjEtNTlhYi00NmE0LWFlOTktZTI3ODc2ZjcwOTY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTAxVDE5MjAzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwNWQ0N2IyZjVkNzNkNzRhOWI4NDhiMDc5MTYzMjg3OGRjZDIzNmZhYmFkZjc5ZDgwODk0OWMwYWI5MTk3ZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.okwdcu6jCcwcMyWIZeYRw7KXkicTZ6zlz2igtqaJ9jg) Subject = Abronia maritima, Object = Isocoma menziesii var. vernonioides
> 
> However, the inverse is not correct
> 
> ![image](https://private-user-images.githubusercontent.com/16688784/439698213-35905269-70bd-44d9-814d-2a7e286f33ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYxMjc1MzEsIm5iZiI6MTc0NjEyNzIzMSwicGF0aCI6Ii8xNjY4ODc4NC80Mzk2OTgyMTMtMzU5MDUyNjktNzBiZC00NGQ5LTgxNGQtMmE3ZTI4NmYzM2NlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTAxVDE5MjAzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhMmY2ZGNjNThiZjFlYWFmZGVhNWZiNTI5ZDE4NDk3MzFiZjg3YTNiYWM5MjA3N2ZlZDNiMGQ0N2RjMWYyYzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.v_Qzli7SwyL7NQDMvjvQdQAHYdYeRAsA0peNHvqIhPs)
> 
> This is saying that subject = Isocoma menziesii var. vernonioides AND object = Isocoma menziesii var. vernonioides.

You should be able to use the same logic that is being used to display the association in the public display page (/collections/individual/index.php):
![image](https://github.com/user-attachments/assets/887979a1-80f3-4b61-8fff-da6426268c77)

themerekat commented:
I'm sure this is happening because it's just grabbing the verbatimSciname from the table
![image](https://github.com/user-attachments/assets/e42ada02-c1d1-4b0f-9ed7-fa09b2d76192)
but in cases when you report the inverse, you need to grab verbatimSciname of omoccurassociations.occid from omoccurrences.sciname dynamically, I guess.

Atticus29 committed:
Merge branch 'Development' into 1924-export-associations-as-resource-relationship

Atticus29 committed:
attempt to resolve off-target sciName issue reported by Katie

Atticus29 commented:
> I'm sure this is happening because it's just grabbing the verbatimSciname from the table ![image](https://private-user-images.githubusercontent.com/16688784/439699067-e42ada02-c1d1-4b0f-9ed7-fa09b2d76192.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1NTA3NTAsIm5iZiI6MTc0NjU1MDQ1MCwicGF0aCI6Ii8xNjY4ODc4NC80Mzk2OTkwNjctZTQyYWRhMDItYzFkMS00YjBmLTllZDctZmEwOWIyZDc2MTkyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA2VDE2NTQxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2ZDJlMzIwYmI4OGE5OGQ4NWVmODZjMmI2OTU2YTcxNWQ5NWQ4YjQzOWVlODgwYTMyYjRjZjcyYzRiNjVmZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.9fdM6FWcCYyb6a3TUfnNVriqK0_13c9geCyASSFEd94) but in cases when you report the inverse, you need to grab verbatimSciname of omoccurassociations.occid from omoccurrences.sciname dynamically, I guess.

@themerekat made an attempt to fix this in ebfe692bc. Does it work now for your case? And importantly, can you tell whether it breaks anything else? I'm no longer certain of anything at all lol, but I can't find any immediate reversions.

Atticus29 committed:
try adding another taxa join

Atticus29 committed:
fix sciName targeting object

Atticus29 commented:
@themerekat j/k last time. Maybe fixed in 9234f1465?

themerekat commented:
Did you touch collections/individual/index.php in this PR? Some linked resources aren't showing up. E.g., 
![image](https://github.com/user-attachments/assets/15b76af1-04b6-4e3e-ab48-c02fce633f67)

![image](https://github.com/user-attachments/assets/b4981823-d29a-48a0-b47d-4a6d8c4cb98d)

Notice that there are linked resources, but nothing on the main page of the record shows this.

(this likely needs a new PR)

Atticus29 commented:
> Did you touch collections/individual/index.php in this PR? Some linked resources aren't showing up. E.g., ![image](https://private-user-images.githubusercontent.com/16688784/441386510-15b76af1-04b6-4e3e-ab48-c02fce633f67.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDI3ODAsIm5iZiI6MTc0NjY0MjQ4MCwicGF0aCI6Ii8xNjY4ODc4NC80NDEzODY1MTAtMTViNzZhZjEtMDRiNi00ZTNlLWFiNDgtYzAyZmNlNjMzZjY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDE4Mjc1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkMzExZTEyM2Y1MDU3YzU3NmJjMDQ2YmY3NmVhMjcwNTJhYjdmNGU1MTMxM2Y3Y2UxZTIxOWU1OGNkMWFiNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XvYwOHLYvOhOImefHQBhubVO3Q7HnOXtyrpFWwiSpN8)
> 
> ![image](https://private-user-images.githubusercontent.com/16688784/441386589-b4981823-d29a-48a0-b47d-4a6d8c4cb98d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDI3ODAsIm5iZiI6MTc0NjY0MjQ4MCwicGF0aCI6Ii8xNjY4ODc4NC80NDEzODY1ODktYjQ5ODE4MjMtZDI5YS00OGEwLWI0N2QtNGE2ZDhjNGNiOThkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDE4MjgwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEwZGZkNDBhZmU4MmUxYzE2NTY1MjE5MzFkZGVkOGUxZjkyNDYyYzBkN2I0MjQyMDVhMWUwOGJjMjMzOGU2ZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AhwJgDzvZlJjJIu9qnToWlguVPJZqxmo1VaKHLQ9KNU)
> 
> Notice that there are linked resources, but nothing on the main page of the record shows this.
> 
> (this likely needs a new PR)

<img width="894" alt="Screenshot 2025-05-07 at 11 54 51 AM" src="https://github.com/user-attachments/assets/3cd6dd8c-64b7-4c50-b7f4-6a0cc4e59950" />

themerekat commented:
> > Did you touch collections/individual/index.php in this PR? Some linked resources aren't showing up. E.g., ![image](https://private-user-images.githubusercontent.com/16688784/441386510-15b76af1-04b6-4e3e-ab48-c02fce633f67.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDI3ODAsIm5iZiI6MTc0NjY0MjQ4MCwicGF0aCI6Ii8xNjY4ODc4NC80NDEzODY1MTAtMTViNzZhZjEtMDRiNi00ZTNlLWFiNDgtYzAyZmNlNjMzZjY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDE4Mjc1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkMzExZTEyM2Y1MDU3YzU3NmJjMDQ2YmY3NmVhMjcwNTJhYjdmNGU1MTMxM2Y3Y2UxZTIxOWU1OGNkMWFiNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XvYwOHLYvOhOImefHQBhubVO3Q7HnOXtyrpFWwiSpN8)
> > ![image](https://private-user-images.githubusercontent.com/16688784/441386589-b4981823-d29a-48a0-b47d-4a6d8c4cb98d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDI3ODAsIm5iZiI6MTc0NjY0MjQ4MCwicGF0aCI6Ii8xNjY4ODc4NC80NDEzODY1ODktYjQ5ODE4MjMtZDI5YS00OGEwLWI0N2QtNGE2ZDhjNGNiOThkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDE4MjgwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEwZGZkNDBhZmU4MmUxYzE2NTY1MjE5MzFkZGVkOGUxZjkyNDYyYzBkN2I0MjQyMDVhMWUwOGJjMjMzOGU2ZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AhwJgDzvZlJjJIu9qnToWlguVPJZqxmo1VaKHLQ9KNU)
> > Notice that there are linked resources, but nothing on the main page of the record shows this.
> > (this likely needs a new PR)
> 
> <img alt="Screenshot 2025-05-07 at 11 54 51 AM" width="894" src="https://private-user-images.githubusercontent.com/2775448/441402426-3cd6dd8c-64b7-4c50-b7f4-6a0cc4e59950.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2NDYyODcsIm5iZiI6MTc0NjY0NTk4NywicGF0aCI6Ii8yNzc1NDQ4LzQ0MTQwMjQyNi0zY2Q2ZGQ4Yy02NGI3LTRjNTAtYjdmNC02YTBjYzRlNTk5NTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MDdUMTkyNjI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWU0NTk4MmUwMDYzMjRmOTA1MWU5ZDJlZjhiNDI1MWRkODIwMDlhY2NmMmRhZmM0YTUxNTFjNDY2NjlkNDI5YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.9mucYPoEe8AIBs5fOu5ac6S0IS6SHZTWydeTjAA8YM4">

Created new issue: https://github.com/Symbiota/Symbiota/issues/2500

themerekat commented on classes/DwcArchiverCore.php:
```diff
@@ -1120,6 +1128,35 @@ private function writeMetaFile(){
+			$extElem3->appendChild($coreIdElem3);
+
+			$mofCnt = 1;
+			foreach ($this->fieldArrMap['associations'] as $term) {
```
Ah, I think you're right. I think calling it resourceRelationship would be more consistent with Darwin Core

Atticus29 replied:
Fixed in 86e426458. Note that I changed it to resourceRelationships to better match the existing camel case and plural pattern (to the extent that there is one).

egbot committed:
DwC meta adjustments

- Switch DwC-TDWG term definitions to Qualname rather than term information link
- Use general DwC terms definitions for scientificName, basisOfRecord, and dynamicProperties, and include these within the DwC schema output type
- Misc minor adjustments

Atticus29 committed:
resolve merge conflicts

Atticus29 committed:
change filename from associations to resourceRelationships

Atticus29 committed:
change to inner join

Atticus29 commented on classes/DwcArchiverResourceRelationship.php:
```diff
@@ -0,0 +1,144 @@
+				// $this->sqlBase = 'SELECT ' . trim($sqlFrag, ', ') . ' FROM omoccurassociations ';
+				$this->sqlBase = 'SELECT DISTINCT ' . trim($sqlFrag, ', ') . ' FROM omoccurrences o
+					INNER JOIN omoccurassociations oa ON o.occid = oa.occidAssociate
+					LEFT JOIN omoccurrences oo ON oo.occid = oa.occid
```
Fixed in 3e711457f.

Atticus29 committed:
remove launch.json

Atticus29 commented:
> Should the /.vscode/launch.json file be removed since it might cause conflicts with other devs using VS for their IDE?

Removed in a2076e43c. Looks like it might be getting tracked in the Development branch as well.
